File: strategy-engine/src/main/java/com/quantstream/strategy/strategies/DonchianChannelStrategy.java
package com.quantstream.strategy.strategies;

import com.quantstream.strategy.framework.TradingStrategy;
import com.quantstream.strategy.model.Signal;
import com.quantstream.strategy.utils.IndicatorUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.stereotype.Component;

import java.time.Instant;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Donchian Channel Strategy
 *
 * Type: Breakout / Trend Following
 *
 * Components:
 * - Upper Channel: Highest high over N periods
 * - Lower Channel: Lowest low over N periods
 * - Middle Channel: (Upper + Lower) / 2
 *
 * Logic:
 * - Price breaks above upper channel → BUY (bullish breakout)
 * - Price breaks below lower channel → SELL (bearish breakout)
 *
 * Key Insight:
 * - Used in Turtle Trading strategy
 * - Captures momentum breakouts from consolidation
 * - Narrow channel → Low volatility → Big move coming
 *
 * Performance:
 * - Best in trending markets (catches breakouts early)
 * - Poor in ranging markets (false breakouts)
 */
@Component
public class DonchianChannelStrategy implements TradingStrategy {

    private static final Logger log = LoggerFactory.getLogger(DonchianChannelStrategy.class);

    @Autowired
    private JdbcTemplate jdbcTemplate;

    @Autowired
    private IndicatorUtils indicators;

    private static final int CHANNEL_PERIOD = 20;

    // Track previous price position relative to channels
    private final Map<String, Boolean> wasAboveUpper = new HashMap<>();
    private final Map<String, Boolean> wasBelowLower = new HashMap<>();

    @Override
    public String getName() {
        return "DONCHIAN_CHANNEL";
    }

    @Override
    public int getRequiredHistoryDays() {
        return CHANNEL_PERIOD;
    }

    @Override
    public Signal analyze(String symbol) {
        try {
            // Step 1: Query historical prices
            List<Double> prices = queryPrices(symbol, CHANNEL_PERIOD);

            // Step 2: Validate data
            if (prices.size() < CHANNEL_PERIOD) {
                log.debug("Not enough data for {}: {} ticks (need {})",
                    symbol, prices.size(), CHANNEL_PERIOD);
                return null;
            }

            // Step 3: Calculate Donchian Channel
            IndicatorUtils.DonchianChannel channel = indicators.calculateDonchianChannel(
                prices, CHANNEL_PERIOD
            );

            double currentPrice = prices.get(0);

            // Step 4: Check price position relative to channels
            boolean isAboveUpper = currentPrice > channel.getUpper();
            boolean isBelowLower = currentPrice < channel.getLower();

            // Step 5: Get previous states
            Boolean prevAboveUpper = wasAboveUpper.get(symbol);
            Boolean prevBelowLower = wasBelowLower.get(symbol);

            // Step 6: Detect breakouts
            Signal signal = null;

            if (prevAboveUpper != null && prevBelowLower != null) {
                // Bullish breakout: Price breaks above upper channel
                if (isAboveUpper && !prevAboveUpper) {
                    signal = new Signal(
                        symbol,
                        "BUY",
                        getName(),
                        calculateConfidence(currentPrice, channel, true),
                        Instant.now()
                    );
                    log.info("Donchian bullish breakout: {} (price={:.2f}, upper={:.2f})",
                        symbol, currentPrice, channel.getUpper());
                }

                // Bearish breakout: Price breaks below lower channel
                else if (isBelowLower && !prevBelowLower) {
                    signal = new Signal(
                        symbol,
                        "SELL",
                        getName(),
                        calculateConfidence(currentPrice, channel, false),
                        Instant.now()
                    );
                    log.info("Donchian bearish breakout: {} (price={:.2f}, lower={:.2f})",
                        symbol, currentPrice, channel.getLower());
                }
            } else {
                log.debug("First run for {}, initializing Donchian state", symbol);
            }

            // Step 7: Store current states
            wasAboveUpper.put(symbol, isAboveUpper);
            wasBelowLower.put(symbol, isBelowLower);

            return signal;

        } catch (Exception e) {
            log.error("Donchian Channel strategy failed for {}: {}", symbol, e.getMessage());
            return null;
        }
    }

    /**
     * Query recent prices from QuestDB.
     */
    private List<Double> queryPrices(String symbol, int limit) {
        String sql = "SELECT price FROM ticks WHERE symbol = ? ORDER BY timestamp DESC LIMIT ?";

        return jdbcTemplate.query(
            sql,
            (rs, rowNum) -> rs.getDouble("price"),
            symbol,
            limit
        );
    }

    /**
     * Calculate confidence based on breakout strength and channel width.
     *
     * Stronger breakout + narrow channel = higher confidence.
     */
    private double calculateConfidence(double price, IndicatorUtils.DonchianChannel channel,
                                      boolean isBuy) {
        double baseConfidence = 0.75;

        double channelWidth = channel.getUpper() - channel.getLower();
        if (channelWidth == 0) return baseConfidence;

        if (isBuy) {
            // Distance above upper channel
            double breakoutStrength = (price - channel.getUpper()) / channelWidth;
            return Math.min(baseConfidence + (breakoutStrength * 0.15), 0.90);
        } else {
            // Distance below lower channel
            double breakoutStrength = (channel.getLower() - price) / channelWidth;
            return Math.min(baseConfidence + (breakoutStrength * 0.15), 0.90);
        }
    }
}

File: backtester/app/strategies/donchian_channel_strategy.py
"""
Donchian Channel Breakout Strategy

Buys when price breaks above upper channel.
Sells when price breaks below lower channel.

Matches Phase 2: DonchianChannelStrategy.java
"""

from typing import Optional
import pandas as pd
from app.strategies.base_strategy import BaseStrategy
from app.models.signal import Signal
from app.core.indicators import calculate_donchian_channel


class DonchianChannelStrategy(BaseStrategy):
    """
    Donchian Channel Breakout Strategy.

    Parameters:
        - Channel Period: 20

    Logic:
        - BUY: Price breaks above upper channel
        - SELL: Price breaks below lower channel

    Confidence:
        - Based on breakout strength and channel width
        - Range: 0.75 to 0.90
    """

    CHANNEL_PERIOD = 20

    def __init__(self):
        super().__init__(name="Donchian Channel")

    def reset_state(self):
        """Reset channel breakout flags."""
        self.was_above_upper = False
        self.was_below_lower = False

    def get_required_candles(self) -> int:
        """Need 20 candles for Donchian Channel."""
        return self.CHANNEL_PERIOD

    def generate_signal(self, df: pd.DataFrame) -> Optional[Signal]:
        """
        Generate signal based on channel breakout.

        Args:
            df: DataFrame with 'high', 'low' prices

        Returns:
            Signal (BUY/SELL) or None
        """
        # Calculate Donchian Channel
        df_with_dc = calculate_donchian_channel(df.copy(), period=self.CHANNEL_PERIOD)

        upper = df_with_dc['dc_upper'].iloc[-1]
        lower = df_with_dc['dc_lower'].iloc[-1]
        current_price = df['close'].iloc[-1]
        signal = None

        # BUY: Price breaks above upper channel
        if current_price > upper and not self.was_above_upper:
            self.was_above_upper = True
            confidence = self._calculate_confidence(current_price, upper, lower, is_buy=True)
            signal = Signal(action="BUY", confidence=confidence)

        # Reset flag when price falls back
        if current_price <= upper:
            self.was_above_upper = False

        # SELL: Price breaks below lower channel
        if current_price < lower and not self.was_below_lower:
            self.was_below_lower = True
            confidence = self._calculate_confidence(current_price, upper, lower, is_buy=False)
            signal = Signal(action="SELL", confidence=confidence)

        # Reset flag when price rises back
        if current_price >= lower:
            self.was_below_lower = False

        return signal

    def _calculate_confidence(self, price: float, upper: float,
                            lower: float, is_buy: bool) -> float:
        """
        Calculate confidence based on breakout strength and channel width.

        Args:
            price: Current price
            upper: Upper channel value
            lower: Lower channel value
            is_buy: True for BUY signal, False for SELL

        Returns:
            Confidence between 0.75 and 0.90
        """
        channel_width = upper - lower
        if channel_width == 0:
            return 0.75

        if is_buy:
            # Distance above upper channel
            breakout_strength = (price - upper) / channel_width
        else:
            # Distance below lower channel
            breakout_strength = (lower - price) / channel_width

        bonus = min(0.15, breakout_strength * 0.5)
        return min(0.90, 0.75 + bonus)
